blaze-server/src/main/scala/org/http4s/server/blaze/Http1ServerStage.scala

```diff
@@ -140,19 +142,21 @@ private[blaze] class Http1ServerStage[F[_]](
     parser.collectMessage(body, requestAttrs) match {
       case Right(req) =>
         executionContext.execute(new Runnable {
-          def run(): Unit =
-            F.runAsync {
-                try serviceFn(req)
-                  .getOrElse(Response.notFound)
-                  .handleErrorWith(serviceErrorHandler(req))
-                catch serviceErrorHandler(req)
-              } {
-                case Right(resp) =>
-                  IO(renderResponse(req, resp, cleanup))
+          def run(): Unit = {
+            val action = optionTSync
+              .suspend(serviceFn(req))
+              .getOrElse(Response.notFound)
+              .recoverWith(serviceErrorHandler(req))
```

Comment:
I think it has to be `recoverWith`, because the error handler is partial, and if it doesn't handle all cases, we mask the original error with a `MatchError`.

We could argue for appending a last resort error handler to the `serviceErrorHandler` in case the user configured something stupid.  And we could argue whether that should handle all `Throwable`, or just `scala.util.NonFatal`, or everything but `VirtualMachineError`, or just `Exception`.  By not appending, we're letting users determine what's fatal, and providing a sensible default error handler.

I'd like to try to render a basic 500 if we haven't successfully rendered anything yet and an error slips through the cracks, but that requires more state tracking.  We've already paid for that on the servlet side (`servletResponse.isCommitted`), but would have to buy it on the blaze side.

Reply:
This is fair, I think :+1: